src/engine/authentication.md

```diff
@@ -0,0 +1,52 @@
+- RPC port exposed towards the browser, allowing malicious webpages to submit messages to the EL engine api. 
+
+The authentication scheme is _not_ designed to 
+
+- Prevent attackers with capability to read ('sniff') network traffic from reading the traffic, 
+- Prevent attackers with capability to read ('sniff') network traffic from performing replay-attacks of earlier messages. 
+
+Authentication is performed as follows: 
+
+- For `HTTP` dialogue, each `jsonrpc` request is individually authenticated by supplying `JWT` token in the HTTP header. 
+- For a WebSocket dialogue, only the initial handshake is authenticated, after which the message dialogue proceeds without further use of JWT.
+  - Clarification: The websocket handshake starts with the client performing a websocket upgrade request. This is a regular http GET request, and the actual
+parameters for the WS-handshake are carried in the http headers. 
+- For `inproc`, a.k.a raw ipc communication, no authentication is required, under the assumption that a process able to access `ipc` channels for the process, which usually means local file access, is already sufficiently permissioned that further authentication requirements do not add security. 
+
+
+## JWT specifications
+
+- Client software MUST expose the authenticated Engine API at a port independent from legacy JSON-RPC API.
+  - The default port for the authenticated Engine API is `8551`. The Engine API is exposed under the `engine` namespace.
+- On the authenticated endpoint, the legacy API **MUST** also be available.
+- The EL **MUST** support at least the following `alg` `HMAC + SHA256` (`HS256`)
+- The EL **MUST** reject the `alg` `none`. 
+
+
+The HMAC algorithm means that symmetric encryption is used, thus several CL's will be able to use the same key, and, from an authentication perspective be able to impersonate each other. From a deployment perspective, it means that an EL does not need to be provisioned with individual keys for each caller. 
+
+## Key distribution
+
+The `EL` and `CL` clients **MUST** accept a cli/config parameter: `jwt-secret`, a `256` bit key, to be used for verifying/generating jwt tokens. 
+If such a parameter is not given, the client **SHOULD** generate such a token, valid for the duration of the execution, and show the token in the output, which the user can then use to provision the counterpart client with.
+
+## JWT Claims
+
+The JWT payload/claims should include: 
+
+- Required: `iat` (issued-at) claim. The EL **SHOULD** only accept `iat` timestamps which are within +-5 seconds from the current time. 
+- Optional: `id` claim. The CL **MAY** use this to communicate a unique identifier for the individual CL node. 
+- Optional: `clv` claim. The CL **MAY** use this to communicate the CL node type/version. 
+
+Other claims included in the JWT payload, which the EL does not handle, **MUST** be ignored. 
```

Comment:
What does "**MUST** be ignored" mean in this context? Must be excluded from the hash computation while checking the signature? I think we need more clarity on what does it mean.

Reply:
I meant that if we later on decided to add more claims, e.g. if we decide that we should also include a paylaod signature, then the CL can do that and if the EL hasn't been updated yet to handle that, things will still work.

Reply:
> Must be excluded from the hash computation while checking the signature? I

No, that's not possible. One nice thing with JWT, as opposed to e.g. SAML, is that signature checking only requires the ability to parse base64. That's it -- no json parsing, no knowledge about the semantics of the claim data.

(As opposed to SAML: where validation requires full XML canonicalization + transformation + path navigation.)

Reply:
I see. So, "ignored" in this context means more claims are permitted, just not specified in the document, right? It sounds reasonable to me.